### PR TITLE
 Add framework for Per-Monitor DPI Awareness on Windows

### PR DESCRIPTION
**Pull request by MaartenBent**
Cherry-picked from #334.

This will make wxWidgets per-monitor DPI aware on Windows when it is marked as `PerMonitorV2` in the manifest.

The remaining changes in #334 include fixes for individual controls.

### CODE REVIEW

File: interface/wx/event.h
Comment (PBfordev) [inline interface/wx/event.h]: The description describes behaviour of "PerMonitorV2", which according to the link below is supported only since Windows 10 version **1703** (Creator's Update). OTOH, V1 which doesn't send the message to child controls and IIRC does not even scale the NC area properly, is available since Windows 8.1....

File: include/wx/msw/toplevel.h
Comment (PBfordev) [inline include/wx/msw/toplevel.h]: Sorry, just being curious: why did you decide to not pass the arguments by const reference here?
I am similarly curious about why some member functions names start with `WX`?

**Comment by PBfordev** [comment]
Thank you very much for working on this!

I haven't had time to try the patch yet but I wonder if this change just works automagically and does not require any changes to the user code to continue working? For example, `wxScrolled<>` with children windows needing to adjust its virtual size after the size of the children changed etc.

File: interface/wx/event.h
Comment (MaartenBent) [reply interface/wx/event.h]: Thanks for noticing. It should indeed be 1703.
I chose to require V2 because V1 (or `true/pm`) misses a lot of functionality. Non-client area scaling was possible in 1607 (you had to enable it manually with `EnableNonClientDpiScaling`) but it did not scale GUI elements like checkboxes and radio buttons.

File: include/wx/msw/toplevel.h
Comment (MaartenBent) [reply include/wx/msw/toplevel.h]: I think I just forgot to use reference.

Do you mean `WXAdjustToPPI`? I think @vadz added it to emphasize that it is an internal function in a public header.

File: include/wx/event.h
Comment (vadz) [inline include/wx/event.h]: Style nitpick: could you please use trailing commas. I'm aware of the benefits of this style but, as often, I think consistency is more important.

File: include/wx/msw/missing.h
Comment (vadz) [inline include/wx/msw/missing.h]: We use `WX_MSW_DECLARE_HANDLE` in `wx/defs.h`, so, again, just for consistency, I'd use it here too.

File: include/wx/msw/missing.h
Comment (vadz) [inline include/wx/msw/missing.h]:
```suggestion
enum DPI_AWARENESS
{
```

File: include/wx/msw/toplevel.h
Comment (vadz) [reply include/wx/msw/toplevel.h]: Yes, this should indeed be
```suggestion
    bool HandleDPIChange(const wxSize& newDPI, const wxRect& newRect);
```

File: include/wx/msw/toplevel.h
Comment (vadz) [inline include/wx/msw/toplevel.h]: I think a slightly more detailed comment here, notably explaining whether `m_activeDPI` is used if we're not per monitor DPI aware and when is it changed, would be useful.

File: src/msw/toplevel.cpp
Comment (vadz) [inline src/msw/toplevel.cpp]: I wonder why do we bother checking for this: AFAICS this can't be called at all in this case as we're simply not going to get `WM_DPICHANGED`. If anything, I'd probably set `m_perMonitorDPIaware` to true here if we do get it, although this shouldn't happen neither in theory.

P.S. After reading the entire patch, I don't think `m_perMonitorDPIaware` is actually used at all... Perhaps we don't even need it (nor the function used for initializing it)?

**Comment by MaartenBent** [comment]
All standard (GDI-based) Windows controls will scale. Other objects, like `wxBitmap`, `wxImage`, `wxDC` (font might scale), `wxGraphicsContext`, `wxGLCanvas` and any other image/drawing related controls are not affected.

I haven't tested `wxScrolled<>` specifically, but if there is a sample using it, I can have a look.

File: src/msw/toplevel.cpp
Comment (vadz) [inline src/msw/toplevel.cpp]: Are the casts to int really necessary? I fail to see what do they add.

I'd probably just write `wxSize(LOWORD(wParam), HIWORD(wParam))` below directly anyhow...

File: src/msw/toplevel.cpp
Comment (vadz) [inline src/msw/toplevel.cpp]: Is this correct? If it is, i.e. if `GetDpiForWindow()` still returns the old DPI, I don't see why do we need to store `m_activeDPI` at all when we can just always retrieve it.

If it isn't, i.e. if `GetDpiForWindow()` already returns the new value here, then we should have initialized `m_activeDPI` before, i.e. it should never be `wxDefaultSize` here.

File: src/msw/toplevel.cpp
Comment (vadz) [inline src/msw/toplevel.cpp]: Maybe it's just me, but I think writing just
```
if ( newDPI != m_activeDPI )
{
    MSWUpdateOnDPIChange(m_activeDPI, newDPI);
    m_activeDPI = newDPI;
}
```
would've been clearer.

**Review by vadz — COMMENTED** [review]
Globally this looks very nice and (or, rather, because of) it's finally so simple -- this is great, thanks a lot once again!

My 2 non-trivial questions are related to MSW behaviour that I'm not sure about:

1. What does `GetDpiForWindow()` return during `WM_DPICHANGED` handling?
2. Can we get `WM_DPICHANGED` without being per-monitor DPI aware in the first place?

Do you already know the answers to those or should we do some more testing?

File: include/wx/msw/toplevel.h
Comment (vadz) [reply include/wx/msw/toplevel.h]: > Do you mean `WXAdjustToPPI`? I think @vadz added it to emphasize that it is an internal function in a public header.

Yes, it's the convention used for non-public functions that are nevertheless accessible in the headers. This will hopefully prevent people from using them and, at the very least, will prevent accidental conflicts with user-defined methods.

**Comment by vadz** [comment]
> All standard (GDI-based) Windows controls will scale. Other objects, like `wxBitmap`, `wxImage`, `wxDC` (font might scale), `wxGraphicsContext`, `wxGLCanvas` and any other image/drawing related controls are not affected.

(see #1470 for the last one)
 
> I haven't tested `wxScrolled<>` specifically, but if there is a sample using it, I can have a look.

`samples/scroll` has a few of those. I think we need to scale the scroll step with DPI too, although this is probably not awfully critical.

**Comment by MaartenBent** [comment]
> What does GetDpiForWindow() return during WM_DPICHANGED handling?

The new DPI.

> Can we get WM_DPICHANGED without being per-monitor DPI aware in the first place?

No, but is is send when running in `<dpiAwareness>PerMonitor</dpiAwareness>` or `<dpiAware>true/pm</dpiAware>` and these are not supported. If someone chooses to use these, I prefer not scaling at all over scaling with lots of artifacts. `m_perMonitorDPIaware` is used to check if `PerMonitorV2` is used.

**Comment by vadz** [comment]
> > What does GetDpiForWindow() return during WM_DPICHANGED handling?
> 
> The new DPI.

Just to be clear, this means that the current code is wrong, doesn't it? Because the comparison  `GetDpi() != newDPI` will always be false then.

> > Can we get WM_DPICHANGED without being per-monitor DPI aware in the first place?
> 
> No, but it is sent when running in `<dpiAwareness>PerMonitor</dpiAwareness>` or `<dpiAware>true/pm</dpiAware>` and these are not supported. If someone chooses to use these, I prefer not scaling at all over scaling with lots of artifacts. `m_perMonitorDPIaware` is used to check if `PerMonitorV2` is used.

I see, thanks for the explanation.

If we need to keep `m_perMonitorDPIaware`, I wonder if we need to store it per-window. It seems like it's always the same for all windows in the application, isn't it? In any case, it's not really a problem to keep the current version, I'm mostly just trying to understand how do things work.

**Comment by PBfordev** [comment]
I had a bit of time to play with this PR using samples on Windows 10 v1903 (build 18362.295), both screens have the same resolution but one was set to 125% DPI. I have noticed several issues (not talking about those for controls using bitmaps or with custom drawing such as `wxGrid`). 

When I launch the minimal sample on the HiDPI screen and move it to the standard DPI screen, the font size for message in `wxStatusBar` field widths seems to be [somewhat different](https://imgur.com/aSSfwPR), then when the sample is launched from the standard DPI screen: a rounding issue? The font actual seems a bit too small for sample started on 100% DPI screen? The moved window also has a different size.

I then checked the widgets sample. There seem to be several issues with `wxRadioBox`. I have also noticed that when I launch the sample from the standard DPI screen instead of the HiDPI one, the text size for most controls is incorrect (even when the sample appears first on the HiDPI screen).

I have not done more testing as I do not know if it would be of any use. Please let me know if I should try more tests as the user (I probably can't be of much help programming-wise) and if so, what would be the best way to report issues - I think where there is too much conversation in the PR, it gets difficult to follow (at least for me).

**Comment by MaartenBent** [comment]
Thanks for testing. In the start post of #334 I added an example manifest (click on `Patch to enable Per-Monitor DPI-Awareness in samples for CMake build system.`) that works fine for me.

> the text size for most controls is incorrect

I see it now too. It doesn't happen when running the application via the debugger (it always starts on correct screen?) but when when running as stand-alone and starting from the non-main montitor I see the problem. I'll commit a fix.

`wxStatusBar` problem could be related. But the text of the status bar is drawn via a `wxDC` which is not targeted in this PR. The status bar position and `wxRadioBox` font are also known problems and already fixed in #334. These are not included in this PR to keep it as simple as possible.

> Is there a reference application, which is (1) using a wide-range of native controls, (2) works as expected with per monitor DPI, and (3) is available for free or as a trial?

I tested #334 with the widgets and dialogs samples. Note that lots of control specific fixes are not included in this PR. Also, some of the already implemented DPI and font changes in master are slightly different from #334 so there could be new bugs.

> ... should be exactly the same as ...

Yes, I belief so.

**Comment by MaartenBent** [comment]
> Just to be clear, this means that the current code is wrong, doesn't it? Because the comparison `GetDpi() != newDPI` will always be false then.
> 

You mean `oldDPI != newDPI`? This is correct. But `if ( oldDPI == wxDefaultSize ) oldDPI = GetDPI();` seems indeed unnecessary, because `oldDPI` (`m_activeDPI`) should always have a valid value. It is initialized after the window has a valid hwnd.

> If we need to keep m_perMonitorDPIaware, I wonder if we need to store it per-window. It seems like it's always the same for all windows in the application, isn't it? In any case, it's not really a problem to keep the current version, I'm mostly just trying to understand how do things work.

Theoretically a window can have a different DPI awareness. Something similar as I use in https://github.com/wxWidgets/wxWidgets/pull/334/commits/90f307f7074e5cf2a7585cdaa55c4812561cb80f can be used to create a frame/window with system DPI.

**Comment by PBfordev** [comment]
Ah I was being and idiot and I did not realize that, unlike a menu or toolbar, `wxStatusBar` is not a native control.

Regarding the manifest, comparing yours (working) and Microsoft's (not working with MSVS 2017 and 2019), the difference is that MS uses http**s** protocol in the URL.
BTW, by a reference application, I meant a 3rd party application, not using wxWidgets: to use it as a reference how things should look. But if the assumption in **Edit 2** in my previous post is true, there is no need for it.

**Comment by vadz** [comment]
> 
> 
> > Just to be clear, this means that the current code is wrong, doesn't it? Because the comparison `GetDpi() != newDPI` will always be false then.
> 
> You mean `oldDPI != newDPI`? This is correct. But `if ( oldDPI == wxDefaultSize ) oldDPI = GetDPI();` seems indeed unnecessary, because `oldDPI` (`m_activeDPI`) should always have a valid value. It is initialized after the window has a valid hwnd.

Yes, this is what I meant, thanks for understanding me correctly, even though I expressed it sloppily.

> > If we need to keep m_perMonitorDPIaware, I wonder if we need to store it per-window. It seems like it's always the same for all windows in the application, isn't it? In any case, it's not really a problem to keep the current version, I'm mostly just trying to understand how do things work.
> 
> Theoretically a window can have a different DPI awareness. Something similar as I use in [90f307f](https://github.com/wxWidgets/wxWidgets/commit/90f307f7074e5cf2a7585cdaa55c4812561cb80f) can be used to create a frame/window with system DPI.

I see, thanks.

**Comment by MaartenBent** [comment]
> Sorry, if I come as only critising: unfortunately with my level of skill all I can do is point out the issues, not fix them. However, if you would like me to nag you about possible problems again, please let me know and I will try my best.

You don't. Please test the latest commit and verify that the font size is now correct.

If it works, I'll add `GetPointSizeAtPPI(int lfHeight, int ppi)` instead of duplicating this 4 times.

File: src/msw/toplevel.cpp
Comment (MaartenBent) [reply src/msw/toplevel.cpp]: It is used, explained in https://github.com/wxWidgets/wxWidgets/pull/1499#issuecomment-524089850.

**Comment by PBfordev** [comment]
The font size of controls now seems to mostly change properly, regardless of which screen was the sample launched from. I did not debug the "screen" issue but now even if I launch the sample from the regular DPI screen (using Total Commander), it always appears on the hiDPI one.

The only difference I spotted after the patch is the button size (testing with those on Gauge page of the widgets sample). The rectangles of buttons for sample launched from the hiDPI screen are taller and wider, the font size seems same, i.e., the gap around the text is bigger for the hiDPI screen. The gap for sample launched from hiDPI seems too large compared to other buttons such as in a message box. The difference persists after moving the frame between screens.

BTW, having PerMonitorV2 results in bunch of issues with standard dialogs (as seen from the dialogs sample), but you are probably aware of this.

**Comment by vadz** [comment]
> The font size of controls now seems to mostly change properly, regardless of which screen was the sample launched from. I did not debug the "screen" issue but now even if I launch the sample from the regular DPI screen (using Total Commander)

I wonder how many Total Commander users are in the wx community. With you and me there are at least 2 of us!

> The only difference I spotted after the patch is the button size (testing with those on Gauge page of the widgets sample). The rectangles of buttons for sample launched from the hiDPI screen are taller and wider, the font size seems same, i.e., the gap around the text is bigger for the hiDPI screen. The gap for sample launched from hiDPI seems too large compared to other buttons such as in a message box. The difference persists after moving the frame between screens.

We definitely can't cache the size in `wxButtonBase::GetDefaultSize()` with per-monitor DPI (or we need to adjust it by DPI ratio) and I don't think this is done yet even in #334 and it's definitely done here. @MaartenBent, do you already maintain a TODO list of remaining tasks for high DPI support? We probably should have one somewhere, I'm just not sure what would be the best place for it.

**Comment by MaartenBent** [comment]
> We definitely can't cache the size in `wxButtonBase::GetDefaultSize()` with per-monitor DPI (or we need to adjust it by DPI ratio) and I don't think this is done yet even in #334 and it's definitely done here.

This function is actually never hit. The main problem for the button size is the cached font size in `wxWindowBase::GetDlgUnitBase`. We discussed it in https://github.com/wxWidgets/wxWidgets/pull/334#issuecomment-450598789 and the following (hidden) comments but there is no fix yet (except removing `static`).

> do you already maintain a TODO list of remaining tasks for high DPI support? We probably should have one somewhere, I'm just not sure what would be the best place for it.

I have a list of controls that still need fixing. I think adding tickets on trac, and one main ticket to track all of them, is the easiest.

**Comment by PBfordev** [comment]
> I have a list of controls that still need fixing. I think adding tickets on trac, and one main ticket to track all of them, is the easiest.

I do realize I am a nobody and should not be that active here but I am an opinionated nobody so I will voice my two cents anyway:

1. I believe that this should not be merged into the master when there are known issues with adverse effects on existing user code.
2. The issue this PR deals with is a very complex and complicated; I think it is beyond a (manageable) scope of a PR.
3. Wouldn't it be better to create a separate branch (I know wxWidgets does not use this model anymore), with each issue dealt with in a separate PR to this branch where new PRs would not be open unless the open one was dealt with and merged into that branch?

**Comment by vadz** [comment]
> I do realize I am a nobody and should not be that active here but I am an opinionated nobody so I will voice my two cents anyway:

Although we accept constructive comments from anybody, you're a valuable contributor to the project, so your comments carry even more weight, and you should voice them, of course.

>     1. I believe that as long as there are known adverse effects on existing user code this should not be merged into the master. 

I'm not sure if we're speaking about actual regressions or the problems with some controls that only manifest when PMv2 awareness is turned on? AFAIU nobody uses PMv2 with wx applications right now because support for it is just not implemented, so any problems when it's turned on don't really count as regressions IMO (besides, if somebody does use this awareness mode, don't they already have these problems anyhow?).

And I'm not aware of any regressions introduced by this PR in non-PMv2 mode, could you please point out if I'm missing/forgetting about something?

> Perhaps it would be OK if it would be documented that some of the advanced custom-drawn and may be less used controls (ribbon, wxDVC...) do not really support PMv2 DPI awareness yet but at least all native and most frequently used controls should work flawlessly.

Eventually, i.e. in 3.2.0, yes, definitely.
 
>     2. The issue this PR deals with is a very complex and complicated; I think it is beyond a (manageable) scope of a PR.

Agreed, which is why Maarten has split #334 into several smaller PRs (and I'm very thankful for this). This particular PR is pretty manageable on its own IMO.
 
>     3. Wouldn't it be better to create a separate branch (I know wxWidgets does not use this model anymore), with each issue addressed in a separate PR to this branch where new PRs would not be open unless the open one was dealt with and merged into that branch?

Maarten does have a branch for this, so the only question is whether this branch should be merged into master at one go at the end or in smaller chunks as it's developed. I definitely prefer the latter because it allows to review the changes -- reviewing the entire set of them in #334 wasn't really feasible.

I do agree that no regressions should be introduced into master, but I just don't see what would merging this one actually break (as opposed to "not make working"). Again, please let me know what am I missing. TIA!

**Comment by PBfordev** [comment]
> I'm not sure if we're speaking about actual regressions or the problems with some controls that only manifest when PMv2 awareness is turned on? AFAIU nobody uses PMv2 with wx applications right now because support for it is just not implemented, so any problems when it's turned on don't really count as regressions IMO (besides, if somebody does use this awareness mode, don't they already have these problems anyhow?).

I was talking about potential PMv2 aware applications only. I did not think that having something in the master not really working that well would be of any real use. And if people turned on PMv2 awareness, they could be surprised that it does not work as expected. While this could be useful for bug reporting, if it is not documented what is not supposed to work, people may spend their time and effort on hunting and reporting issues the maintainers are already well aware of... 

TLDR: I believe that if this gets merged "incomplete", its state should be publicly documented (what works and what does not) so the people are not surprised at the possible breakage when turning PMv2 on in their apps.

**Comment by vadz** [comment]
The hope is that the rest of #334 will be integrated relatively quickly, so most things should work. Until then we should indeed mention that PMv2 is WIP and not supposed to work yet. @MaartenBent do you agree?

**Comment by MaartenBent** [comment]
I agree. Merging in small batches is a lot easier.

The only changes in this PR that could cause regressions are the `WXAdjustToPPI` and `GetNonClientMetrics` changes. I have to make sure these work correctly in no/high dpiAware. All other code is not used when PMv2 is not enabled.

In addition, I can add a `wxSystemOptions` `msw.window.pmdpi` that (for now) has to be enabled for PMv2 to work. Once the pending changes are merged, I can change its behavior to be enabled by default.

**Comment by vadz** [comment]
> In addition, I can add a `wxSystemOptions` `msw.window.pmdpi` that (for now) has to be enabled for PMv2 to work. Once the pending changes are merged, I can change its behavior to be enabled by default.

I don't think it's worth adding a new system option. Just a note in the documentation saying that per-monitor DPI support is an experimental feature not supposed to work just yet is enough IMO.

**Comment by MaartenBent** [comment]
I added a note to the `wxDPIChangedEvent` documentation.

I tested with no and high DPI awareness and found no font regressions. IMO this PR can be merged.

File: src/msw/menuitem.cpp
Comment (vadz) [inline src/msw/menuitem.cpp]: Hmm, shouldn't we add `wxWindow* window = NULL` to `wxNativeFontInfo` ctor to let it do the right thing immediately? Having to remember to adjust the point size afterwards looks much more error prone.

In fact, I'd even say that we should add `wxWindow* window` argument, i.e. without `= NULL`, to force explicitly specifying the null window if we really don't have anything to give it.

Of course, this doesn't need to be part of this PR, but it still looks like a good idea, what do you think?

**Review by vadz — APPROVED** [review]
Thanks again! I'll merge this once the CI builds finish (I don't see why would they fail, but experience shows that it's always better to wait to avoid any surprises...).

File: src/msw/menuitem.cpp
Comment (MaartenBent) [reply src/msw/menuitem.cpp]: I can add it to this PR.

But `wxNativeFontInfo` constructor can be used publicly so it might break user code if there is no default argument. What do you prefer?

Comment (vadz) [reply src/msw/menuitem.cpp]: Oh, I didn't think about this, but you're right, even though this function was never supposed to be public, undoubtedly somebody did use it somewhere... We could use an ugly hack with providing `= NULL` only when not building the library (`WXBUILDING` is not defined), I guess, or just always have the default argument and rely on the discipline inside the library.

But, again, you're right and it's almost certainly not a good idea to use it unconditionally, sorry for the bad advice.

File: interface/wx/event.h
Comment (PBfordev) [inline interface/wx/event.h]: Do the methods need to be tagged as "since" - the whole class is already marked as such...

File: interface/wx/event.h
Comment (PBfordev) [inline interface/wx/event.h]: This may be just nitpicking, but I would not a word resolution here. I would say that when speakign about displays, resolution is something like 1920x1080 (i.e., the number of pixels a display has) and this does not change with a DPI change. Microsoft uses [Display Scale Factor](https://docs.microsoft.com/en-us/windows/win32/hidpi/high-dpi-desktop-application-development-on-windows), others use pixel density...

File: interface/wx/event.h
Comment (MaartenBent) [reply interface/wx/event.h]: Yes, they are probably unnecessary. `wxGestureEvent` also does not have them.

**Comment by vadz** [comment]
I'm merging this even though further enhancements are still possible -- just to allow @MaartenBent to continue with cherry-picking the remaining parts of #334. Thanks again!

**Pull request merged**
